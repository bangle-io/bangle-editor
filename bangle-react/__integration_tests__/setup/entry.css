@import 'bangle-core/style.css';
