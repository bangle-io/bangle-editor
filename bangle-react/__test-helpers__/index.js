export * from './pjsx';
export * from './react-test-editor';
