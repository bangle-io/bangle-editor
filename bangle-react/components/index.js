import * as dinos from './dinos/index';

export * as dinos from './dinos/index';

export const spec = (opts = {}) => {
  return [dinos.spec(opts.dinos)];
};

export const plugins = (opts = {}) => {
  return [dinos.plugins(opts.dinos)];
};

export const commands = {
  dinos: dinos.commands,
};
