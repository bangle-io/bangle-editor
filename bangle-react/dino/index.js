export * as dino from './dino';
