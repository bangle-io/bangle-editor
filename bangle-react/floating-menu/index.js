export * as floatingMenu from './menu';
