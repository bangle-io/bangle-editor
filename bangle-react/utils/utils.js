export function createElement(type = 'div', attrs = {}) {
  const element = document.createElement(type);
  for (let attr in attrs) {
    const value = attrs[attr];
    if (attr.startsWith('data-')) {
      element.setAttribute(attr, value);
    } else if (attr === 'id') {
      element.id = value;
    } else if (attr === 'class') {
      value.split(' ').forEach((c) => element.classList.add(c));
    } else {
      throw new Error(`Attr ${attr} not supported`);
    }
  }
  return element;
}
