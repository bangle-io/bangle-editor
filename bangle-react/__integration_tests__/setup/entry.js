import './entry.css';
import React from 'react';
import reactDOM from 'react-dom';
import { corePlugins, coreSpec } from 'bangle-core/index';
import { SpecSheet } from 'bangle-core/spec-sheet';
import { Slice } from 'prosemirror-model';
import { DOMSerializer } from 'prosemirror-model';
import { dino } from 'bangle-react/dino';
import { stopwatch } from 'bangle-react/stopwatch';
import * as prosemirrorView from 'prosemirror-view';
import { ReactEditor } from 'bangle-react/ReactEditor';

window.prosemirrorView = prosemirrorView;

console.debug('Bangle-react entry.js');
setup();

function setup() {
  const element = document.createElement('div');
  window.document.body.appendChild(element);
  element.setAttribute('id', 'root');
  // const editorContainer = document.createElement('div');
  // editorContainer.setAttribute('id', 'pm-root');
  // element.appendChild(editorContainer);
  // setupEditor(editorContainer);
  reactDOM.render(<App />, element);
}

function App() {
  const specSheet = new SpecSheet([
    ...coreSpec(),
    stopwatch.spec(),
    dino.spec(),
  ]);
  const plugins = [...corePlugins(), stopwatch.plugins(), dino.plugins()];
  window.commands = {
    stopwatch: stopwatch.commands,
    dino: dino.commands,
  };
  const onEditorReady = (_editor) => {
    window.editor = _editor;
  };
  const renderNodeViews = ({ node, ...args }) => {
    if (node.type.name === 'dino') {
      return <dino.Dino node={node} {...args} />;
    }
  };
  window.dispatcher = (command) => {
    return command(
      window.editor.view.state,
      window.editor.view.dispatch,
      window.editor.view,
    );
  };
  return (
    <ReactEditor
      options={{ id: 'pm-root', specSheet, plugins }}
      onReady={onEditorReady}
      renderNodeViews={renderNodeViews}
    />
  );
}

const createEvent = (name, options = {}) => {
  let event;
  if (options.bubbles === undefined) {
    options.bubbles = true;
  }
  if (options.cancelable === undefined) {
    options.cancelable = true;
  }
  if (options.composed === undefined) {
    options.composed = true;
  }
  event = new Event(name, options);

  return event;
};

window.manualPaste = function manualPaste(htmlStr) {
  function sliceSingleNode(slice) {
    return slice.openStart === 0 &&
      slice.openEnd === 0 &&
      slice.content.childCount === 1
      ? slice.content.firstChild
      : null;
  }

  function doPaste(view, text, html, e) {
    let slice = prosemirrorView.__parseFromClipboard(
      view,
      text,
      html,
      view.shiftKey,
      view.state.selection.$from,
    );

    if (
      view.someProp('handlePaste', (f) => f(view, e, slice || Slice.empty)) ||
      !slice
    ) {
      return;
    }

    let singleNode = sliceSingleNode(slice);
    let tr = singleNode
      ? view.state.tr.replaceSelectionWith(singleNode, view.shiftKey)
      : view.state.tr.replaceSelection(slice);
    view.dispatch(
      tr.scrollIntoView().setMeta('paste', true).setMeta('uiEvent', 'paste'),
    );
  }
  const pasteEvent = createEvent('paste');

  doPaste(window.editor.view, null, htmlStr, pasteEvent);
};

window.domSerializer = function domSerializer() {
  return DOMSerializer.fromSchema(window.editor.view.state.schema);
};
