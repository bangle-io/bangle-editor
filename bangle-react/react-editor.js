import React from 'react';
import reactDOM from 'react-dom';
import PropTypes from 'prop-types';
import { objUid } from 'bangle-core/utils/object-uid';
import { BangleEditor } from 'bangle-core/editor';
import { saveRenderHandlers } from 'bangle-core/node-view';
import { bangleWarn } from 'bangle-core/utils/js-utils';

const LOG = false;

let log = LOG ? console.log.bind(console, 'react-editor') : () => {};

const nodeViewUpdateCallbackCache = new WeakMap();

export class ReactEditor extends React.PureComponent {
  static propTypes = {
    options: PropTypes.object.isRequired,
    renderNodeViews: PropTypes.func,
    onReady: PropTypes.func,
  };

  editorRenderTarget = React.createRef();
  state = { nodeViews: [] };

  renderHandlers = {
    create: (nodeView, nodeViewProps) => {
      log('create');
      this.setState({
        nodeViews: [...this.state.nodeViews, nodeView],
      });
    },
    update: (nodeView, nodeViewProps) => {
      log('update');
      const updateCallback = nodeViewUpdateCallbackCache.get(nodeView);
      // I am thinking that this might be called before react had the chance
      // to mount. I believe saving of args in nodeViewPropsCache will get the react
      // render up to date with latest nodeViewProps
      if (updateCallback) {
        updateCallback();
      }
    },
    destroy: (nodeView) => {
      log('destroy');
      this.setState({
        nodeViews: this.state.nodeViews.filter((n) => n !== nodeView),
      });
    },
  };

  async componentDidMount() {
    const { options } = this.props;
    // save the renderHandlers in the dom to decouple nodeView instantiating code
    // from the editor. Since PM passing view when nodeView is created, the author
    // of the component can get the handler reference from `getRenderHandlers(view)`.
    // Note: this assumes that the pm's dom is the direct child of `editorRenderTarget`.
    saveRenderHandlers(this.editorRenderTarget.current, this.renderHandlers);
    this.editor = new BangleEditor(this.editorRenderTarget.current, options);
    if (this.props.onReady) {
      this.props.onReady(this.editor);
    }
  }

  componentWillUnmount() {
    this.editor.destroy();
    this.editor = null;
  }

  render() {
    log('rendering PMEditorWrapper');
    return (
      <>
        <div
          ref={this.editorRenderTarget}
          id={this.props.options.id}
          data-testid={this.props.options.testId}
        />
        {this.state.nodeViews.map((nodeView) => {
          return reactDOM.createPortal(
            <NodeViewElement
              nodeView={nodeView}
              renderNodeViews={this.props.renderNodeViews}
            />,
            nodeView.mountDOM,
            objUid.get(nodeView),
          );
        })}
      </>
    );
  }
}

class NodeViewElement extends React.PureComponent {
  update = () => {
    this.setState({ nodeViewProps: this.props.nodeView.getNodeViewProps() });
  };

  constructor(props) {
    super(props);
    // So that we can directly update the nodeView without the mess
    // of prop forwarding. This is okay because a nodeView and ReactComponent has
    // 1:1 mapping always.
    nodeViewUpdateCallbackCache.set(props.nodeView, this.update);
    this.state = { nodeViewProps: this.props.nodeView.getNodeViewProps() };
  }

  attachToContentDOM = (reactElement) => {
    if (!reactElement) {
      return;
    }
    const { contentDOM } = this.props.nodeView;
    // Since we do not control how many times this callback is called
    // make sure it is not already mounted.
    if (!reactElement.contains(contentDOM)) {
      // If contentDOM happens to be mounted to someone else
      // remove it from there.
      if (contentDOM.parentNode) {
        contentDOM.parentNode.removeChild(contentDOM);
      }
      reactElement.appendChild(contentDOM);
    }
  };

  getChildren() {
    if (!this.props.nodeView.contentDOM) {
      return null;
    }

    if (this.state.nodeViewProps.node.isInline) {
      return (
        <span className="bangle-content-mount" ref={this.attachToContentDOM} />
      );
    }

    return (
      <div className="bangle-content-mount" ref={this.attachToContentDOM} />
    );
  }

  render() {
    const element = this.props.renderNodeViews({
      ...this.state.nodeViewProps,
      children: this.getChildren(),
    });
    if (!element) {
      bangleWarn(
        'renderNodeView prop must return a react element for the node',
        this.state.nodeViewProps.node,
      );
      throw new Error(
        `Missing react render for node of type "${this.state.nodeViewProps.node.type.name}"`,
      );
    }
    return element;
  }
}
