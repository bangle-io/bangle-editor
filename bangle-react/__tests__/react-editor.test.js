/** @jsx pjsx */
/**
 * @jest-environment jsdom
 */

import { render, fireEvent } from '@testing-library/react';
import { BangleEditor, corePlugins, coreSpec } from 'bangle-core/index';
import { SpecSheet } from 'bangle-core/spec-sheet';
import { ReactEditor } from 'bangle-react/react-editor';
import { getRenderHandlers, NodeView } from 'bangle-core/node-view';
import { safeInsert, removeSelectedNode } from 'bangle-core/utils/pm-utils';
import { bananaComponent, Banana } from './setup/banana';
import { pjsx } from 'bangle-react/__test-helpers__';
import { sleep } from 'bangle-core/utils/js-utils';
import { selectNodeAt } from 'bangle-core/test-helpers/index';
import { Node } from 'prosemirror-model';
import { EditorView } from 'prosemirror-view';
const consoleError = console.error;

beforeEach(() => {
  console.error = consoleError;
});

const insertBananaAtSelection = (attrs = {}) => (dispatch, state) => {
  const node = state.schema.nodes.banana.create(attrs);
  const newTr = safeInsert(node, state.selection.from)(state.tr);
  if (dispatch) {
    dispatch(newTr);
  }
  return true;
};

describe('basic tests', () => {
  let specSheet, plugins;
  beforeEach(() => {
    specSheet = new SpecSheet([...coreSpec()]);
    plugins = [...corePlugins()];
  });

  test('mounts correctly', async () => {
    const onReady = jest.fn();
    const result = await render(
      <ReactEditor
        options={{
          id: 'test',
          specSheet,
          plugins,
        }}
        onReady={onReady}
      />,
    );

    expect(onReady).toBeCalledTimes(1);
    expect(onReady).toHaveBeenNthCalledWith(1, expect.any(BangleEditor));
    expect(result.container).toMatchInlineSnapshot(`
      <div>
        <div
          id="test"
        >
          <div
            class="ProseMirror bangle-editor"
            contenteditable="true"
          >
            <p>
              <br />
            </p>
          </div>
        </div>
      </div>
    `);
  });

  test('Unmounting destroys editor', async () => {
    let editor;
    const onReady = (_editor) => {
      editor = _editor;
    };
    const result = await render(
      <ReactEditor
        options={{
          id: 'test',
          specSheet,
          plugins,
        }}
        onReady={onReady}
      />,
    );
    expect(editor.view.docView).toBeTruthy();
    await result.unmount();
    expect(editor.view.docView).toBeNull();
  });
});

describe('rendering node views', () => {
  test('persists node view render handlers', async () => {
    let editor;
    const onReady = (_editor) => {
      editor = _editor;
    };
    await render(
      <ReactEditor
        options={{
          id: 'test',
          specSheet: new SpecSheet([...coreSpec()]),
          plugins: [...corePlugins()],
        }}
        onReady={onReady}
      />,
    );

    expect(getRenderHandlers(editor.view)).toEqual({
      create: expect.any(Function),
      update: expect.any(Function),
      destroy: expect.any(Function),
    });
  });

  test('Throws error if rendering is not implemented', async () => {
    console.error = jest.fn();
    let editor;
    const onReady = (_editor) => {
      editor = _editor;
    };

    const banana = bananaComponent('react-editor-test');
    const spec = new SpecSheet([...coreSpec(), banana.spec()]);
    expect(() =>
      render(
        <ReactEditor
          options={{
            id: 'test',
            specSheet: spec,
            plugins: [...corePlugins(), banana.plugins()],
            stateOpts: {
              doc: (<doc>
                <para>
                  hello <banana color="brown" /> world
                </para>
              </doc>)(spec.schema),
            },
          }}
          onReady={onReady}
          renderNodeViews={() => {}}
        />,
      ),
    ).toThrowErrorMatchingInlineSnapshot(
      `"renderNodeView must handle rendering for node of type \\"banana\\""`,
    );
  });

  test('Handles add and delete of node views', async () => {
    console.error = jest.fn();
    let editor;
    const onReady = (_editor) => {
      editor = _editor;
    };

    const renderNodeViews = jest.fn(({ node, ...args }) => {
      if (node.type.name === 'banana') {
        return <Banana node={node} {...args} />;
      }
      throw new Error('Unknown node');
    });

    const banana = bananaComponent('react-editor-test');
    const spec = new SpecSheet([...coreSpec(), banana.spec()]);

    const { container } = await render(
      <ReactEditor
        options={{
          id: 'test',
          specSheet: spec,
          plugins: [...corePlugins(), banana.plugins()],
          stateOpts: {
            doc: (<doc>
              <para>hello world</para>
            </doc>)(spec.schema),
          },
        }}
        onReady={onReady}
        renderNodeViews={renderNodeViews}
      />,
    );
    const view = editor.view;
    insertBananaAtSelection()(view.dispatch, view.state);
    await sleep(10);

    expect(view.state).toEqualDocAndSelection(
      <doc>
        <para>
          <banana />
          hello world
        </para>
      </doc>,
    );

    insertBananaAtSelection({ color: 'pink' })(view.dispatch, view.state);
    insertBananaAtSelection({ color: 'brown' })(view.dispatch, view.state);

    expect(renderNodeViews).toBeCalledTimes(3);

    await sleep(10);

    expect(view.state).toEqualDocAndSelection(
      <doc>
        <para>
          <banana color={'brown'} />
          <banana color={'pink'} />
          <banana color={'yellow'} />
          hello world
        </para>
      </doc>,
    );

    expect(container).toMatchSnapshot();

    // Select pink one
    selectNodeAt(view, 2);
    expect(view.state.selection.node.attrs).toMatchObject({ color: 'pink' });

    view.dispatch(removeSelectedNode(view.state.tr));

    expect(renderNodeViews).toBeCalledTimes(4);

    expect(view.state).toEqualDocAndSelection(
      <doc>
        <para>
          <banana color={'brown'} />
          <banana color={'yellow'} />
          hello world
        </para>
      </doc>,
    );

    expect(container).toMatchSnapshot();
  });

  test('handles updating of attrs', async () => {
    console.error = jest.fn();
    let editor;
    const onReady = (_editor) => {
      editor = _editor;
    };

    const renderNodeViews = jest.fn(({ node, ...args }) => {
      if (node.type.name === 'banana') {
        return <Banana node={node} {...args} />;
      }
      throw new Error('Unknown node');
    });

    const banana = bananaComponent('react-editor-test');
    const spec = new SpecSheet([...coreSpec(), banana.spec()]);

    const { container } = await render(
      <ReactEditor
        options={{
          id: 'test',
          specSheet: spec,
          plugins: [...corePlugins(), banana.plugins()],
          stateOpts: {
            doc: (<doc>
              <para>hello world</para>
            </doc>)(spec.schema),
          },
        }}
        onReady={onReady}
        renderNodeViews={renderNodeViews}
      />,
    );
    const view = editor.view;
    insertBananaAtSelection({ color: 'blue' })(view.dispatch, view.state);
    insertBananaAtSelection({ color: 'red' })(view.dispatch, view.state);
    insertBananaAtSelection({ color: 'orange' })(view.dispatch, view.state);
    await sleep(10);

    const orangeBananaDOM = container.querySelector(`[data-color="orange"]`);
    expect(orangeBananaDOM).toMatchInlineSnapshot(`
      <div
        data-color="orange"
      >
        I am fresh orange banana
      </div>
    `);

    fireEvent.click(orangeBananaDOM);
    await sleep(10);
    expect(orangeBananaDOM).toMatchInlineSnapshot(`
      <div
        data-color="orange"
      >
        I am stale orange banana
      </div>
    `);

    expect(view.state).toEqualDocAndSelection(
      <doc>
        <para>
          <banana ripe="stale" color={'orange'} />
          <banana color={'red'} />
          <banana color={'blue'} />
          hello world
        </para>
      </doc>,
    );
  });

  test('Calls rendering with correct params', async () => {
    console.error = jest.fn();
    let editor;
    const onReady = (_editor) => {
      editor = _editor;
    };

    const renderNodeViews = jest.fn(({ node, ...args }) => {
      if (node.type.name === 'banana') {
        return <Banana node={node} {...args} />;
      }
      throw new Error('Unknown node');
    });

    const banana = bananaComponent('react-editor-test');
    const spec = new SpecSheet([...coreSpec(), banana.spec()]);

    const { container } = await render(
      <ReactEditor
        options={{
          id: 'test',
          specSheet: spec,
          plugins: [...corePlugins(), banana.plugins()],
          stateOpts: {
            doc: (<doc>
              <para>hello world</para>
            </doc>)(spec.schema),
          },
        }}
        onReady={onReady}
        renderNodeViews={renderNodeViews}
      />,
    );
    const view = editor.view;
    insertBananaAtSelection({ color: 'blue' })(view.dispatch, view.state);

    expect(renderNodeViews).toBeCalledTimes(1);
    expect(renderNodeViews).toHaveBeenNthCalledWith(1, {
      node: expect.any(Node),
      view: expect.any(EditorView),
      getPos: expect.any(Function),
      decorations: expect.anything(),
      selected: expect.any(Boolean),
      attrs: {
        color: 'blue',
        ripe: 'fresh',
      },
      updateAttrs: expect.any(Function),
      children: null,
    });

    const orangeBananaDOM = container.querySelector(`[data-color="blue"]`);

    fireEvent.click(orangeBananaDOM);
    await sleep(10);

    expect(renderNodeViews).toBeCalledTimes(2);
    expect(renderNodeViews).toHaveBeenNthCalledWith(2, {
      node: expect.any(Node),
      view: expect.any(EditorView),
      getPos: expect.any(Function),
      decorations: expect.anything(),
      selected: expect.any(Boolean),
      attrs: {
        color: 'blue',
        ripe: 'stale',
      },
      updateAttrs: expect.any(Function),
      children: null,
    });
  });
});
