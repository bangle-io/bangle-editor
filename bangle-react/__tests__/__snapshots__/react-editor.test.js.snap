// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering node views Handles add and delete of node views 1`] = `
<div>
  <div
    id="test"
  >
    <div
      class="ProseMirror bangle-editor"
      contenteditable="true"
    >
      <p>
        <span
          data-bangle-container=""
          data-testid="react-editor-test"
          draggable="true"
        >
          <div
            data-color="brown"
          >
            I am fresh brown banana
          </div>
        </span>
        <span
          data-bangle-container=""
          data-testid="react-editor-test"
          draggable="true"
        >
          <div
            data-color="pink"
          >
            I am fresh pink banana
          </div>
        </span>
        <span
          data-bangle-container=""
          data-testid="react-editor-test"
          draggable="true"
        >
          <div
            data-color="yellow"
          >
            I am fresh yellow banana
          </div>
        </span>
        hello world
      </p>
    </div>
  </div>
</div>
`;

exports[`rendering node views Handles add and delete of node views 2`] = `
<div>
  <div
    id="test"
  >
    <div
      class="ProseMirror bangle-editor"
      contenteditable="true"
    >
      <p>
        <span
          data-bangle-container=""
          data-testid="react-editor-test"
          draggable="true"
        >
          <div
            data-color="brown"
          >
            I am fresh brown banana
          </div>
        </span>
        <span
          data-bangle-container=""
          data-testid="react-editor-test"
          draggable="true"
        >
          <div
            data-color="yellow"
          >
            I am fresh yellow banana
          </div>
        </span>
        hello world
      </p>
    </div>
  </div>
</div>
`;
